flink-runtime/src/main/java/org/apache/flink/runtime/state/SnapshotDirectory.java

```diff
@@ -0,0 +1,172 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.runtime.state;
+
+import org.apache.flink.core.fs.FileStatus;
+import org.apache.flink.core.fs.FileSystem;
+import org.apache.flink.core.fs.Path;
+
+import javax.annotation.Nonnull;
+
+import java.io.IOException;
+import java.util.concurrent.atomic.AtomicReference;
+
+/**
+ * This class represents a directory that is the target for a state snapshot. This class provides some method that
+ * simplify resource management when dealing with such directories, e.g. it can produce a {@link DirectoryStateHandle}
+ * when the snapshot is completed and disposal considers whether or not a snapshot was already completed. For a
+ * completed snapshot, the ownership for cleanup is transferred to the created directory state handle. For incomplete
+ * snapshots, calling {@link #deleteIfIncompleteSnapshot()} will delete the underlying directory resource.
+ */
+public class SnapshotDirectory {
+
+	/**
+	 * Lifecycle stages of a snapshot directory.
+	 */
+	enum State {
+		ONGOING, COMPLETED, DELETED
+	}
+
+	/** This path describes the underlying directory for the snapshot. */
+	@Nonnull
+	private final Path directory;
+
+	/** The filesystem that contains the snapshot directory. */
+	@Nonnull
+	private final FileSystem fileSystem;
```

Comment:
Do we need `FileSystem` if we have `directory` which is of type `Path`?

Reply:
In this class, I would keep it because it is used in several operations and obtaining the `FileSystem` from `Path` looks like it might not be a cheap operation. So this is more like a cache, or in case that you already have the `FileSystem` when constructing the object.